pkg/compactor/executor.go

```diff
@@ -0,0 +1,325 @@
+	schedulerClient          schedulerpb.CompactorSchedulerClient
+	schedulerConn            *grpc.ClientConn
+	invalidClusterValidation *prometheus.CounterVec
+}
+
+func newSchedulerExecutor(cfg Config, logger log.Logger, invalidClusterValidation *prometheus.CounterVec) (*schedulerExecutor, error) {
+
+	executor := &schedulerExecutor{
+		cfg:                      cfg,
+		logger:                   logger,
+		invalidClusterValidation: invalidClusterValidation,
+	}
+
+	// Initialize scheduler client. This call will succeed even if scheduler is unreachable
+	// since grpc.Dial() creates the connection immediately and connects lazily.
+	var err error
+	executor.schedulerClient, executor.schedulerConn, err = executor.makeSchedulerClient()
+	if err != nil {
+		return nil, err
+	}
+
+	return executor, nil
+}
+
+func (e *schedulerExecutor) run(ctx context.Context, c *MultitenantCompactor) error {
+	workerID := fmt.Sprintf("compactor-%s", c.ringLifecycler.GetInstanceID())
+	level.Info(e.logger).Log("msg", "compactor running in scheduler mode", "scheduler_endpoint", e.cfg.SchedulerAddress, "worker_id", workerID)
+
+	b := backoff.New(ctx, backoff.Config{
+		MinBackoff: e.cfg.SchedulerMinLeasingBackoff,
+		MaxBackoff: e.cfg.SchedulerMaxLeasingBackoff,
+	})
+
+	for {
+		work, err := e.leaseAndExecuteJob(ctx, c, workerID)
+		if err != nil {
+			level.Warn(e.logger).Log("msg", "failed to lease or execute job", "err", err)
+		}
+
+		if work {
+			b.Reset()
+		}
+
+		select {
+		case <-time.After(b.NextDelay()):
+			continue
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (e *schedulerExecutor) stop() error {
+	if e.schedulerConn != nil {
+		return e.schedulerConn.Close()
+	}
+	return nil
+}
+
+// startJobStatusUpdater starts a goroutine that sends periodic IN_PROGRESS keep-alive updates
+func (e *schedulerExecutor) startJobStatusUpdater(ctx context.Context, key *schedulerpb.JobKey, spec *schedulerpb.JobSpec) {
+	ticker := time.NewTicker(e.cfg.SchedulerUpdateInterval)
+	defer ticker.Stop()
+
+	jobId := key.Id
+	jobTenant := spec.Tenant
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := e.updateJobStatus(ctx, key, spec, schedulerpb.IN_PROGRESS); err != nil {
+				level.Warn(e.logger).Log("msg", "failed to send keep-alive update", "job_id", jobId, "tenant", jobTenant, "err", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// sendFinalJobStatus sends a final status update to the scheduler with retry policy.
+// Compaction jobs send final statuses on completion, planning jobs only on failure for reassignment.
+func (e *schedulerExecutor) sendFinalJobStatus(ctx context.Context, key *schedulerpb.JobKey, spec *schedulerpb.JobSpec, status schedulerpb.UpdateType) {
+	jobId := key.Id
+	jobTenant := spec.Tenant
+
+	var err error
+	switch spec.JobType {
+	case schedulerpb.COMPACTION:
+		req := &schedulerpb.UpdateCompactionJobRequest{Key: key, Tenant: spec.Tenant, Update: status}
+		_, err = failsafe.Get(func() (*schedulerpb.UpdateJobResponse, error) {
```

Comment:
Since we're not returning a result from the Failsafe call, we can use `.Run` instead of `.Get`:

```go
	err := failsafe.Run(func() error {
		_, err := e.schedulerClient.PlannedJobs(ctx, req)
		return err
	}, retryPolicy)
```